driver/src/orderbook/streamed/balance.rs

```diff
@@ -0,0 +1,199 @@
+        self.sell_buy_internal(amount, batch_id, Operation::Sell)
+    }
+
+    pub fn buy(&mut self, amount: u128, batch_id: BatchId) -> Result<(), Error> {
+        self.sell_buy_internal(amount, batch_id, Operation::Buy)
+    }
+
+    pub fn revert_sell(&mut self, amount: u128, batch_id: BatchId) -> Result<(), Error> {
+        self.revert_sell_buy_internal(amount, batch_id, Operation::Sell)
+    }
+
+    pub fn revert_buy(&mut self, amount: u128, batch_id: BatchId) -> Result<(), Error> {
+        self.revert_sell_buy_internal(amount, batch_id, Operation::Buy)
+    }
+
+    fn get_balance_internal(
+        &self,
+        batch_id: BatchId,
+        include_withdraw: bool,
+    ) -> Result<U256, Error> {
+        let mut balance = self.balance;
+        if self.deposit.batch_id < batch_id {
+            balance = balance
+                .checked_add(self.deposit.amount)
+                .ok_or(Error::MathOverflow)?;
+        }
+        // TODO: this could temporarily fail while not all trades for solution have been received.
+        if self.pending_proceeds.batch_id < batch_id {
+            balance = balance
+                .checked_add(self.pending_proceeds.increase)
+                .ok_or(Error::MathOverflow)?
+                .checked_sub(self.pending_proceeds.decrease)
+                .ok_or(Error::MathOverflow)?;
+        }
+        if include_withdraw && self.withdraw.batch_id < batch_id {
+            // Saturating because withdraw requests can be for amounts larger than balance.
+            balance = balance.saturating_sub(self.withdraw.amount);
+        }
+        Ok(balance)
+    }
+
+    fn clear_pending_deposit_and_solution_if_possible(&mut self, batch_id: BatchId) {
+        if self.deposit.batch_id < batch_id {
+            self.deposit.amount = 0.into();
+        }
+        if self.pending_proceeds.batch_id < batch_id {
+            self.pending_proceeds.increase = 0.into();
+            self.pending_proceeds.decrease = 0.into();
+        }
+    }
+
+    fn sell_buy_internal(
+        &mut self,
+        amount: u128,
+        batch_id: BatchId,
+        operation: Operation,
+    ) -> Result<(), Error> {
+        match self.pending_proceeds.batch_id.cmp(&batch_id) {
+            Ordering::Less => {
+                let new_balance = self.get_balance_internal(batch_id, false)?;
+                self.balance = new_balance;
+                self.clear_pending_deposit_and_solution_if_possible(batch_id);
+                self.pending_proceeds.batch_id = batch_id;
+                *self.pending_proceeds.get_field(operation) = amount.into();
+            }
+            Ordering::Equal => {
+                let field = self.pending_proceeds.get_field(operation);
+                *field = field
+                    .checked_add(amount.into())
+                    .ok_or(Error::MathOverflow)?
+            }
+            Ordering::Greater => return Err(Error::TradeForPastBatch),
```

Comment:
We never reset the batch_id. Is this not a problem when events are reverted due to reorgs?